src/CoreWCF.Primitives/src/CoreWCF/Channels/BinaryMessageEncoderFactory.cs

```diff
@@ -538,20 +536,20 @@ private ArraySegment<byte> ExtractSessionInformationFromMessage(ArraySegment<byt
                         }
 
                         int size = dictionarySize;
-                        int offset = messageData.Offset + headerSize;
+                        int offset = headerSize;
 
                         while (size > 0)
                         {
                             decoder.Reset();
-                            int bytesDecoded = decoder.Decode(buffer, offset, size);
+                            int bytesDecoded = decoder.Decode(messageData.Slice(offset, size));
                             int utf8ValueSize = decoder.Value;
```

Comment:
Inside of the `if (dictionarySize > 0) block, you could initially slice messageData to (0, headerSize) so it only contains the dictionary header. Then you don't need to do any booking of size  as you read each dictionary item as you can just do `Slice(bytesDecoded)`.

Reply:
I realized headerSize would be the wrong thing, I think it would be `Slice(headerSize, headerSize + DictionarySize) so just give the dictionary. I hope you get what I'm saying,

Reply:
I see it. implemented https://github.com/CoreWCF/CoreWCF/pull/801/commits/652e27f4d056acbe8836bfa2eb4ebb7635c0dd76